src/mem/heap/mspace_malloc.c

```diff
@@ -23,41 +23,100 @@
 
 #include <util/dlist.h>
 #include <util/array.h>
+#include <util/log.h>
 
 #include <kernel/printk.h>
 #include <kernel/panic.h>
 
+#include <mem/heap/mspace_malloc.h>
+
 /* TODO make it per task field */
 //static DLIST_DEFINE(task_mem_segments);
 
 //#define DEBUG
 
 extern struct page_allocator *__heap_pgallocator;
 extern struct page_allocator *__heap_pgallocator2 __attribute__((weak));
-static struct page_allocator ** const mm_page_allocs[] = {
-	&__heap_pgallocator,
-	&__heap_pgallocator2,
+extern struct page_allocator *__heap_fixed_pgallocator __attribute__((weak));
+
+struct mm_heap_allocator {
+	struct page_allocator **pg_allocator;
+	heap_type_t type;
 };
 
+static struct mm_heap_allocator const mm_page_allocs[] = {
+	{ &__heap_pgallocator, HEAP_RAM },
+	{ &__heap_pgallocator2, HEAP_FAST_RAM },
+	{ &__heap_fixed_pgallocator, HEAP_EXTERN_MEM },
+};
+
+static struct mm_heap_allocator const *mm_cur_allocator =
+	&mm_page_allocs[0];
+
 static void *mm_segment_alloc(int page_cnt) {
 	void *ret;
 	int i;
+
```

Comment:
I think irq lock is necesasry here (`mm_cur_allocator` may change in the middle of the function, this may lead to bugs).